### ISSUE TITLE
Completely on-demand services are hard to represent in stop_times.txt

### ISSUE THREAD

**Issue by gcamp** [MobilityData/gtfs-flex#54]
This is my biggest problem currently with the v2 Flex proposition. 

I understand the need to integrate in `stop_times` for service that includes a fixed section. However, for a completely flexible service it adds a lot of strange complexity for a service that will most likely be considered completely differently by consumers. 

Disadvantages that I see 
- It forces creating "trips" that have no meaning on a trip basis. 
- It forces ordering of "stops" even if there are no actual ordering 
- (Relatively subjective) It's hard to understand what a on demand service by looking at the `stop_times`, especially if there are more than 2 or 3 zones

Proposed solution : 
- Have a separated `on_demand_locations.geojson` that contains only on-demand services. And define the relationship between those locations inside the `properties` (details TBD). 
- Remove `location_groups.txt` since it was used mainly for groups of on demand zone (I might be missing something here)

It's quite possible my proposed solution is not up to some of the requirements needed as I haven't spend as much time thinking about the problem as other people here. 

Edit : Concrete proposal on how to improve things in [this post](https://github.com/MobilityData/gtfs-flex/issues/54#issuecomment-906434794)

**Comment by westontrillium** [comment]
> It forces creating "trips" that have no meaning on a trip basis.

Could you explain more? To me, even if "trip" is somewhat inaccurate semantically, the function of a trip in the context of a GTFS dataset is entirely useful for flexible services.

> It forces ordering of "stops" even if there are no actual ordering

The `stop_sequence` field is irrelevant for certain flexible services, but not all (e.g. a deviating route). It doesn't, however, get in the way of representing an on-demand service with multiple stop_time entries. Verifying if a consumer should ignore the `stop_sequence` field is as simple as checking if the entry in question has the `start/end_pickup_dropoff_window` field defined.

> (Relatively subjective) It's hard to understand what a on demand service by looking at the stop_times, especially if there are more than 2 or 3 zones

What about stop_times makes these services more difficult to understand? I do think services with complex zoning are always going to be more of a challenge to figure out by looking at the raw data than the average fixed route service (regardless of where those service parameters are described in the data), however, the information represented in stop_times for a typical single area demand-responsive service can be quickly parsed out.

> Have a separated `on_demand_locations.geojson` that contains only on-demand services. And define the relationship between those locations inside the properties (details TBD).

Interesting. What would be the benefit of isolating locations that are for on-demand services in this way?

> Remove location_groups.txt since it was used mainly for groups of on demand zone (I might be missing something here)

`location_groups.txt` is needed for the following:
- unordered stops services ("I can pickup/drop-off at any of these fixed stops in no particular order between x and y time")
- demand-responsive/on-demand service areas that have a mix of both zones and fixed stops with the same service parameters (there are many cases in which a service will drop the rider off within a town or at a specific fixed location that falls outside that area)
- multiple isolated zones treated as the same service area (i.e. riders can pick up/drop off within **and** between  each zone).
- creates a simpler, less labor intensive avenue for representing more complex service areas with isolated polygons. Instead of creating a single location with multiple isolated polygons, a data producer can bundle multiple area-based locations to describe new service areas. This is especially useful when those isolated polygons already exist elsewhere in the data as contiguous areas for other services.

**Comment by gcamp** [comment]
>Verifying if a consumer should ignore the stop_sequence field is as simple as checking if the entry in question has the start/end_pickup_dropoff_window field defined.

To me that's part of the problem, it's not that it doesn't work, it's that it adds complexity for no reason. Put yourself in the place of a producer creating their first GTFS-flex without knowledge of GTFS, having to understand the concept of `stop_sequence` makes little sense. I need to look more into the OTP implementation but in practice I don't see how a on-demand and fixed service (flexible or not) can reuse code and processing. Why reuse concepts just to have checks for on-demand vs fixed every time? 

> I do think services with complex zoning are always going to be more of a challenge to figure out by looking at the raw data than the average fixed route service

The way I see it, if properties are defined correctly it would be relatively trivial to understand the transfers between zones. You look for zone X and then look for the possible transfers. Compare that to read the entire stop_times for trips with zone X, when trips.txt might not be sorted. 

> location_groups.txt is needed for the following: [...]

Same as previous comment, to me all those requirements could be included in the `geojson` while reducing complexity. 

I'll create an example and post here to make my proposal more clear and concrete.

**Comment by e-lo** [comment]
I am in general agreement with @gcamp and wonder if (relatively) recent evolution in already adopted GTFS Schedule would actually handle most of the issues...and some proposals for GTFS Schedule. Ideally we aren't treating "flexible transit" as a special case where we have to learn/apply/handle/code multiple methods for representing similar things. 

---

One specific thought is to use `stops.txt` to represent every and any pick-up/drop off location point; (and `locations.geojson` to represent zones/polygons).  Many of the groupings in `location_groups.txt` could be represented by the parent station relationship structure.  

From @westontrillium on need for `location_groups.txt`
> unordered stops services

From @gcamp:
> Have a separated on_demand_locations.geojson that contains only on-demand services.

I'm not sure why `locations.geojson` and `stops.txt` wouldn't suffice?  I'm wary of adding more files that we need to encode and super wary of having providers need to keep track of anything in a geojson properties which I anticipate being more difficult (or just frustrating) for them to manage/code/troubleshoot.   

From @westontrillium on need for `location_groups.txt`:
> multiple isolated zones treated as the same service area 

Should be handled using a modification to the proposal within [GTFS Restrictions](https://docs.google.com/document/d/1E0RdoFTNKWEGVjRiaXM92Q1W1p-FbPm_KVt-cZOD4s8/edit#heading=h.is3jpknlubwz). 

For the purpose of fares, this can be handled with a `zone_id`

From @westontrillium on need for `location_groups.txt`: 
> Creates a simpler, less labor intensive avenue for representing more complex service areas with isolated polygons. Instead of creating a single location with multiple isolated polygons, a data producer can bundle multiple area-based locations to describe new service areas. 

Would it be possible to add a use case to the google doc to demonstrate this?  I am having trouble understanding how this would make it simpler but I think concrete use cases will probably highlight.

**Comment by e-lo** [comment]
I do think that `stop_times.txt` is useful for services that have actual stops they serve (with potentially flexible timetable or order) and mixed services. 

I also support @gcamp 's thought that it is overkill and adds complexity for service that is completely flexible w/in a certain zone.  

That said, I'd like to understand if I'm correct in understanding that a completely flexible service within a single location zone would have just a single entry in `stop_times.txt`?

**Comment by scmcca** [comment]
FWIW

@gcamp 
> It forces creating "trips" that have no meaning on a trip basis.

For completely flexible service (i.e., no fixed direction, can travel anywhere within a zone), its true that the concept of `route` and `trip` change slightly. In scheduled transit, a route is serviced by multiple trips throughout the day and has a name, color, branding, etc. 

For flexible service in GOFS (extension of GTFS-Flex), `routes` and `trips` become more of a place where attributes related to flexible service live. For example, `route_type` keeps its meaning, and proposed fields like `trip_type`, `luggage_allowed`, `service_animal_allowed` tell us about rider information and allowances of the "trip" as a service. To me the difference is more of a broadening of concept rather than a misuse of the existing concept.

> It forces ordering of "stops" even if there are no actual ordering

For those in the GOFS working group, we recently added a section in the GOFS Phase 1 doc about directional and non-directional service patterns (point 6 under "Modelization Issues & Suggested solutions"). 

The proposed principle is: 
- Non-directional service patterns (i.e., completely flexible) exist in a single `stop_time` record. 
- Directional service patterns have 2 or more `stop_time` records, and their directionality is defined by ascending `stop_sequence` numbers. 
- These can be combined to have separate flexible zones existing in their respective single `stop_time` records, but directionality between these flexible zones described by `stop_sequence`. See #46 for more. 
- As per #55 , `.start_pickup_dropoff_window` and `.end_pickup_dropoff_window` can be used as an indicator as to whether a given `stop_time` record is flexible.

This is where I am misaligned with @westontrillium that "`stop_sequence` field is irrelevant for certain flexible services". IMHO we should not mutate the GTFS Schedule definition of `stop_sequence`. If we maintain the principle that single `stop_time` records describe non-directional service patterns, and 2 or more `stop_time` records ordered by `stop_sequence` describe directional service patterns we can avoid a lot of confusion and complexity between different service pattern cases. We are seeing this especially as we move forward with GOFS.

> (Relatively subjective) It's hard to understand what a on demand service by looking at the stop_times, especially if there are more than 2 or 3 zones.

This comment can mean a lot of things. In GOFS, `wait_rules.txt` hints at redefining attributes in a sub-zones (i.e., for completely flexible, non-directional service). I imagine that going forward we may broaden `wait_rules.txt` to something like `location_attributes.txt` to be able to define "at one level deeper" other attributes than just wait times, such as pickup/dropoff windows, for different zones in a non-directional service pattern. 

More on this in the GOFS Phase 1 doc, including how to deal with spatial and temporal overlap of geometries with attributes, but here is a diagram that communicates the modelling for multiple zones within a single `stop_time` record:

![Copy of GOFS refactor proposal (draft 2)](https://user-images.githubusercontent.com/62957380/129597417-c512878c-5354-4dac-a2ec-60b6eb13dfc7.png)

I hope the principles I communicated here are clear. On MobilityData's side this is what we brainstormed to make the most sense rationalizing different service patterns. Let me know if I missed something, concerns, or if anything is unclear.

**Comment by westontrillium** [comment]
[Replying to @e-lo]

> One specific thought is to use `stops.txt` to represent every and any pick-up/drop off location point; (and `locations.geojson` to represent zones/polygons). Many of the groupings in `location_groups.txt` could be represented by the parent station relationship structure.

I don't think using `parent_station` would work for a couple reasons off the top of my head:
- Since a stop cannot have more than one `parent_station`, individual locations that are part of more than one service iteration could not be represented (you couldn't have a single stop/location belong to more than one `parent_station`).
- There is an implied hierarchy with using `parent_station` that does not implicitly exist in a set of demand-responsive zones, and no existing `location_type` fits the functional definition of a zone-based stop.

> That said, I'd like to understand if I'm correct in understanding that a completely flexible service within a single location zone would have just a single entry in `stop_times.txt`?

Yes, that's correct.

> Should be handled using a modification to the proposal within GTFS Restrictions.

You'd probably be interested in the conversation over at [#46](https://github.com/MobilityData/gtfs-flex/issues/46). I don't see how TravelRestrictions fits the specific case of describing a multi-zone single service, though.

> Would it be possible to add a use case to the google doc to demonstrate this? I am having trouble understanding how this would make it simpler but I think concrete use cases will probably highlight.

If you're referring to the one that links from bit.ly/gtfs-flex-v2, I'm not sure of the status of that doc and I do not have editing access. For now though, I'll include a real-life example here:

Consider the following Dial-A-Ride ADA service from Brockton Area Transit:
(travel allowed within and between towns)
M-F
6:00am-6:20am
- Brockton, Avon, Abington, Whitman, Stoughton

6:20am-5:50pm
- Brockton, Avon, abington, Whitman, Stoughton, Easton

5:50pm-6:00pm
- Brockton, Avon, Abington, Whitman, Easton

6:00pm-9:05pm
- Brockton, Avon, Abington, Whitman

With location_groups, I only need to build 6 polygons (for each town) and then just group them appropriately using a `location_group_id` within `location_groups.txt` for each span of time (4 `location_group_id`s total).

Without location_groups, I would need to build the 6 polygons representing each town, and then build a polygon for the serviceable area for each span of time (4 more polygons representing the combined service areas). **That's 6 simple polygons vs 10 simple _and_ complex polygons (6 simple and 4 complex).**

location_groups allows a provider to quite easily group various already-existing locations together to describe timespans of shared service as needed.

location_groups would be all the more necessary if @scmcca's approach with `stop_sequence` were codified, as there would be no way to describe services in which travel is permitted between and within two isolated zones in any order.

**Comment by gcamp** [comment]
I think those are good points @westontrillium, see responses below 
> would we also need to have it refer to foreign keys from (eventual) extension data

Yes, but I don't see it as problematic. Yes indeed some keys will be duplicated but some would not which I think would be great for separation of concerns. 

>How would your proposal describe a service where the vehicle picks up riders in zoneA only from 08:00:00 to 09:00:00 and then shuttles them to zoneB dropping off only from 09:30:00 to 10:30:00 without pickup_type/drop_off_type and stop_sequence (or something equivalent)?

This would indeed not be possible in my current proposal. However, I have a hard time seeing how a producer would require something like this. Do you a have real-world example in mind? That said, adding more granular fields as discussed in https://github.com/MobilityData/gtfs-flex/issues/62 would also solve this issue.

> What about route deviation?

I would not change anything for flexible routes, only for on-demand services. Maybe I'm missing something on how those two types of services would interact?  

> If we add fields to on_demand_rules.json to account for the above questions, would that not in some way be adding a similar level of complexity that already exists in Flex's current iteration?

I don't think so but I can be proven wrong. Just needing to understand what would a on-demand trip with >2 stops trumps adding more fields that are simple to understand in terms of complexity in my opinion.

**Comment by e-lo** [comment]
Re json vs csv: since you are for the most part expecting the same data in each entry, I'd vote for keeping it in "csv land" for the sake of consistency.

**Comment by e-lo** [comment]
Re `locations.geojson` + `location_groups.txt` + `stops.txt` all defining items which could be used in `stop_times.txt`

I would _love_ for all the primary keys for locations to exist in a single file for the sake of:
 - ease of validation
 - error-reduction, 
 - lookup-ability (you don't have to look 3 places), and 
 - being able to specify the spec in a (fairly) commonly-accepted specification format.

I think this is feasible....
1. It seems like any information we have in `locations.geojson` _properties_ is repeating `stops.txt` info and could be represented by `stops.txt` ...plus a linkage to the more detailed geography (in `locations.geojson`? (please confirm).
2. `location_groups.txt` is defining relationships which can't be defined in a single row in `stops.txt` with `parent_station`, which makes sense; similar to locations.geojson, we might want to be able to assign some of the same properties to this location group which exist in `stops.txt`. Why not specify the location group _within_ `stops.txt` along with all its properties and then have a separate file which can define more complex relationship structure...but not additional properties (this is basically as is currently specified...just having it also represented in `stops.txt`?

**Comment by e-lo** [comment]
I also have a use case to crack w.r.t. how something like the [Metro Micro](https://micro.metro.net/) service would be coded.

The **rider** sees a continuous service area before booking the travel (for now) with individual stop locations obfuscated:
![image](https://user-images.githubusercontent.com/794787/131606850-9a92cf33-dc57-44eb-9d8a-040dfd9084bf.png)

When booking, it becomes clear that there are actually specific stops which are served (its not door to door)...although it does even tell you "which one" (just "one of"):
<img width="1106" alt="Select a Ride" src="https://user-images.githubusercontent.com/794787/131606984-aebc3c99-aece-4a84-aef6-49cd6d488850.png">

1. I believe that as of now would want to specify this by delineating each possible stopping location in `stops.txt` and then creating an entry for each stopping location in `location_groups.txt` which then referenced a `location_group_id` akin to `metro-micro-glendale-zone`?
2. But then how do we know if/how to present it as a continuous polygon....(and is that desired)?

**Comment by gcamp** [comment]
On CSV : 👍

On location and ids : we could have only stops and reference the geometry in `locations.geojson`. However I'm unsure about which properties we want to assign on a group outside of the group itself. Also I'll add that I said in the past that `location_groups.txt` might not be useful but I changed opinion on this and would keep it. 

On MicroMetro : that's virtual stops. I think within the goal of flex that is mainly describing the service and helping discoverability and I don't think virtual stops are required for that. Adding virtual stops would have several complexities including the fact that you can be redirected to different ones depending on the system in real time and it's not something that can be implemented in a static file. At Transit our micro-transit implantation doesn't take virtual stops into account and it hasn't caused problem. I would also argue that's probably in the GOFS realm.

**Comment by tsherlockcraig** [comment]
> How would your proposal describe a service where the vehicle picks up riders in zoneA only from 08:00:00 to 09:00:00 and then shuttles them to zoneB dropping off only from 09:30:00 to 10:30:00 without pickup_type/drop_off_type and stop_sequence (or something equivalent)?

I think this is a type of service that was described to me by @t2gran in Norway. A sort of "demand response" intercity service where there are pickups in one town, travel along a highway, drop off in another town. Although i don't think it was pickup only in the first and drop off only in the second necessarily.

> flexible routes, only for on-demand services. Maybe I'm missing something on how those two types of services would interact?

There are a lot of services that blend these two modes, for example https://www.riderct.org/client_media/files/routes/shopping/RCT-Map.Kingdom-Shopper.B.8.28.20.pdf which operates as a DRT service between the out and back runs.

Going back to the issue title and the core problem we're trying to solve: what if we just had a simple walkthrough document for "you want to describe a simple general public demand response service in GTFS-flex v2? Easy-->it's just these fields and here's the example data." In other words, the problem we're discussing exists in part just because the GTFS-flex documentation describes the full spec, most of which is not needed for one very common and very simple use case. What if we just built a view of the spec for that simple use case instead of further changing the spec? Have we tried that and would it provide a satisfactory/easier/as effective solution?

**Comment by westontrillium** [comment]
I agree that presenting a shorter overview as @tsherlockcraig describes would demonstrate the simplicity of the spec better. [Here's my attempt at such a document](https://docs.google.com/document/d/1zkH0acut6PIf76jQQDUt-HMkVsqTp9nt_KL4q4TKmis/edit?usp=sharing).

Attached below is also the actual data example (cleaned up to **only** include what is required for this particular service): 
[Flex_simple_on-demand.zip](https://github.com/MobilityData/gtfs-flex/files/7094715/Flex_simple_on-demand.zip)

**Comment by mcplanner** [comment]
What's the purpose of the polygon in that example? If it's illustrative,
but not tied to booking rules, that would be different than if it actually
related to where you could be picked up or dropped off? I could imagine
agencies wanting to do both, but am curious what we're grounding this in
now.

**Comment by scmcca** [comment]
### Data examples gdoc
@westontrillium @e-lo 
>If you're referring to the one that links from bit.ly/gtfs-flex-v2, I'm not sure of the status of that doc and I do not have editing access.

Some of these examples are outdated at this point, particularly the route deviation examples. We are working on updating them to most recent thinking. All reference information should be consulted in this repository as the official documentation, so any examples will be transferred and the gdoc deprecated.

### Response to `on_demand_rules.txt` proposal
@gcamp The model you provided is certainly easier to comprehend, but in doing so is a simplification of different service patterns we are trying to represent. 
- What would be the role of `stop_times.txt` (Required) in a dataset containing `on_demand_rules.txt`? Would we write rules to conditionally forbid `stop_times.txt`? This departure seems to equally entail some problems for validators, which we are all trying to avoid.
- To echo @westontrillium's [example](https://github.com/MobilityData/gtfs-flex/issues/54#issuecomment-906844763), the model you proposed doesn't seem to support overlapping pickup/dropoff windows between zones with different pickup/dropoff rules. Instead, the model you proposed defines directionality (e.g., pickup in `from_id`, dropoff in `to_id`) between zones using `on_demand_rules.from_id` and `.to_id` but defines the pickup/dropoff window across 2 zones (or for 1 zone if `.from_id` = `.to_id`). This misses some granularity as was discussed further. As you said, perhaps #62 addresses this concern?
- RE route deviation: our most recent thinking on handling these cases builds off of the directionality principles in stop_times.txt elaborated in #46 and in a [previous comment](https://github.com/MobilityData/gtfs-flex/issues/54#issuecomment-899653857). Example:
     - What is route deviation? When a fixed schedule service allows a rider to request pickup/dropoff within a deviation area from the fixed route.
     - Using our directionality principles, where non-directional travel is represented in a single `stop_times.txt` record, and directional travel is represented across `stop_times.txt` records with ascending `stop_sequence` values, we would represent route deviation like this:

**stop_times.txt**
| trip_id | start_pickup_dropoff_window | end_pickup_dropoff_window | stop_id | stop_sequence | pickup_type | drop_off_type |
| ----- | ----- | ----- | ----- | ----- | ----- | ----- | 
| trip1 | 10:00 | 10:00 | stop1 | 1 | 0 | 0 |
| trip1 | 10:00 | 10:10 | devationArea1 | 1 | 1 | 3 |
| trip1 | 10:10 | 10:20 | stop2 | 1 | 0 | 0 |
| trip1 | 10:20 | 10:30 | devationArea2 | 1 | 1 | 3 |
| trip1 | 10:30 | 10:40 | stop3 | 1 | 0 | 0 |
| trip1 | 10:40 | 10:50 | devationArea3 | 1 | 1 | 3 |

The table above describes a directionality along the fixed path of stops, with drop off deviation areas between fixed stops that must be requested by the rider. Therefore, its possible that no deviations are requested and the vehicle continues to travel along its fixed path. 

Note that I did not use `arrival_time` or `departure_time` for fixed stops. This was to accomodate a scenario where a vehicle would have a window to arrive at fixed stops based on potential deviations. Ultimately these time fields are up to the agency's policy. Also note that the current spec only allows `drop_off_type=3` for linestrings, this should be changed to accommodate our most recent thinking on route deviation based on directionality principles.

Thoughts?

### ID referencing locations
@e-lo
> I would love for all the primary keys for locations to exist in a single file

Thinking out loud / possible tangent: this reminds me of the solution we arrived at for [Fares v2](https://docs.google.com/document/d/19j-f-wZ5C_kYXmkLBye1g42U-kvfSVgYLkkG5oyBauY/edit#heading=h.mgfbltbi0ht1) using `areas.txt` as the single ID referencing point and grouping mechanism for `stop_times.txt` and `stops.txt` (which was proposed by Transit and I believe is currently in their implementation, cc: @jsteelz). 

The use case in Fares is to group spatial stops (`stops.txt`) and spatiotemporal stops (`stop_times.txt`) to assign to fare prices. However this is a separate and unusable concept in GTFS-Flex. 

In GTFS-Flex we are defining spatial locations (`stops.txt` and `locations.geojson`), grouping them (`location_groups.txt`), and assigning them to directional and temporal information in `stop_times.txt`. 

> It seems like any information we have in locations.geojson properties is repeating stops.txt info and could be represented by stops.txt ...plus a linkage to the more detailed geography (in locations.geojson? (please confirm).

Where in `stops.txt` would the more detailed geography live? An immediate consideration is that `lat` and `lon` (describing a single point) are conditionally required for `location_types=0, 1, 2`. 

A potential solution would be to create a `location_type=5` for "flexible location" and add a field `stops.flexible_location_id` that would reference the ID of a `locations.geojson`. This would mean that consumers would only have to look into `locations.geojson` based on the "activation switch" of `location_type=5` instead of looking into a multiple separate files for every `stop_id` referenced in every record of `stop_times.txt`. This concept could also be used for a `location_type=6` for "flexible location group" that would specify the added `stops.flexible_location_id` field to reference `location_groups.location_group_id`.

While this solution does not do away with looking into separate files, it may reduce the necessity of doing so and save computation for fixed stop transit services. 

TLDR 
- In the current model `stop_times.stop_id` has to check 3 separate files for every record.
- In the proposed model `stop_times.stop_id` need only look to `stops.stop_id`, but `stops.txt` may need to check into either `locations.geojson` or `location_groups.txt` based on the indicator provided in `stops.location_type` (instead of for every record).

@e-lo I'm curious to hear what you had in mind in terms of a specific proposal.

@westontrillium Along these lines, is there a reason why `point` and `multipoint` are not allowed in `locations.geojson`? It seems we could doaway with the need to reference `stops.stop_id` for point locations in `location_groups.txt`. Unfortunately this exposes another concern: For completely flexible service using only polygons, what happens to the Required `stops.txt` file? This is actually a supporting argument for my proposal above suggesting that `stops.txt` be the directory file for finding location IDs.

Thoughts on all the above?

**Comment by westontrillium** [comment]
> To echo @westontrillium's example, the model you proposed doesn't seem to support overlapping pickup/dropoff windows between zones with different pickup/dropoff rules. Instead, the model you proposed defines directionality (e.g., pickup in `from_id`, dropoff in `to_id`) between zones using `on_demand_rules.from_id` and `.to_id` but defines the pickup/dropoff window across 2 zones (or for 1 zone if `.from_id` = `.to_id`). This misses some granularity as was discussed further. As you said, perhaps Clarify `start_pickup_dropoff_window`/`end_pickup_dropoff_window` interpretation #62 addresses this concern?

#62 does address this concern, however I found a case embedded in my [example](https://github.com/MobilityData/gtfs-flex/issues/54#issuecomment-906844763) that I think requires a separate `start_drop_off_window` field if we went the on_demand_rules.txt route. See my [comment](https://github.com/MobilityData/gtfs-flex/issues/62#issuecomment-912765819) in #62 for details.

What still **isn't** addressed is the case in [this comment](https://github.com/MobilityData/gtfs-flex/issues/54#issuecomment-907367676) ("straddling" pickup/drop-offs in separate zones with sequential windows).

@scmcca I agree with your model of Route deviation, except:

> Also note that the current spec only allows `drop_off_type=3` for linestrings, this should be changed to accommodate our most recent thinking on route deviation based on directionality principles.

Route deviation would use a polygon for the deviation zones, though, not linestring (that would be more akin to continuous stops, so you could have `pickup/drop_off_type=2`.

> @westontrillium Along these lines, is there a reason why `point` and `multipoint` are not allowed in `locations.geojson`? It seems we could doaway with the need to reference `stops.stop_id` for point locations in `location_groups.txt`.

I was not involved in the decision making regarding disallowing those types, but I would think it's to avoid having more than one value representing a single stop or point (i.e. you shouldn't need to have two data for one feature). A stop could be served by multiple services, fixed route or otherwise, so it is useful to be able to grab from the stops.txt bag to represent, for example, an on-demand stops service where many/all stops are already represented in the data because of their presence in other routes.

> For completely flexible service using only polygons, what happens to the Required `stops.txt` file?

Well, right now, we include an empty `stops.txt` file in our exports to comply with the spec. Down the road, I could envision amending GTFS so that `stops.txt` is a conditionally required file (if a `locations.geojson` is not present).

**Comment by gcamp** [comment]
>  A sort of "demand response" intercity service where there are pickups in one town, travel along a highway, drop off in another town.

That would not require a different time for dropoff time. The travel distance does the job by itself to create the time between the two stops, the producer doesn't need to specify a different dropoff time. It would be necessary if for some reason the travel time between zone A and B would be 20 minutes but pickup starts at 9:00 with dropoff starting only at 11:00 (but why would anyone do that?). This was actually what was discussed in https://github.com/MobilityData/gtfs-flex/issues/62 and everybody seems to agree the drop-off start time is not useful so I'm not sure if that would be different here. 

> There are a lot of services that blend these two modes, for example

Maybe we have different definition of but to me that seems like  "simple" flexible service. There's indeed a on-demand (pickup_type=2) aspect, but my proposition would only affect service that don't have any fixed service associated to them like micro-transit. (Again, wouldn't change anything how flexible services are represented like the one linked here).


>What if we just built a view of the spec for that simple use case instead of further changing the spec?

By "view of the spec", I'm guessing you mean a API guide or better documentation? That would certainly help but I don't think it would solve most issues. My point is to reduce complexity and the implementation that result from it. Documentation helps but will never change the fact that interpreting the data sometimes requires you to use your [system 2](https://en.wikipedia.org/wiki/Thinking,_Fast_and_Slow#Two_systems) and really think like we did [here](https://gist.github.com/gcamp/c0ca3ae950d2bd7a0bf286ffd8bd0888). 

Comments @scmcca 

> What would be the role of stop_times.txt (Required) in a dataset containing on_demand_rules.txt? Would we write rules to conditionally forbid stop_times.txt? This departure seems to equally entail some problems for validators, which we are all trying to avoid.

The fact that new data model is in a different file is actually a benefit to me. That means that you could create a GTFS with fixed service and `on_demand` service that could be read by backend that don't know about GTFS-flex (which wasn't the case in the existing proposition). Indeed that might require us to make stop_times.txt conditionally required instead of simply required for GTFS feed that only contains on_demand services. 

> the model you proposed doesn't seem to support overlapping pickup/dropoff 

I would repeat my first comment on requesting real-world example of how that would be useful. I can see how #62 is useful by giving a more detailed interpretation of how end of service is represented, but actual overlapping pickup/dropoff is still something I don't see the usefulness. 

> RE route deviation

My proposal would remove `*_window` from `stop_times.txt` (I see now that's not explicitly mentioned, sorry about that). However, nothing else would change on the current proposal. The route deviation examples in the repo would still be compatible with what I suggest. 

> Note that I did not use arrival_time or departure_time for fixed stops. This was to accomodate a scenario where a vehicle would have a window to arrive at fixed stops based on potential deviations. Ultimately these time fields are up to the agency's policy.

I never saw an example that used `*_window` to define implicit uncertainty about a vehicle arrival. To me that seems like twisting the spec to add a concept that it wasn't meant for. I don't think `*_window` was added with that use case in mind, but feel free to contradict me. 

Arrival uncertainty wasn't a concern of GTFS before flexible service, I'm not sure why that would be one now. Would a fixed service allowed to use `*_window` instead of `arrival_time`/`departure_time` entirely? That doesn't seems like something that's desirable. 

That said, I would also request a real-world example of something like this. If those values embedded in the GTFS is not shown in the user at any point, I don't think it should be in the GTFS.

A reminder that as a principle [speculative features are discouraged ](https://github.com/google/transit/blob/master/gtfs/CHANGES.md#speculative-features-are-discouraged). Maybe we should agree on a set of real-world example that we could base the need of the spec on.

**Comment by westontrillium** [comment]
from @gcamp:
> I never saw an example that used `*_window` to define implicit uncertainty about a vehicle arrival. To me that seems like twisting the spec to add a concept that it wasn't meant for. I don't think `*_window` was added with that use case in mind, but feel free to contradict me.

@gcamp Just a quick note that I **do not** think `*_window` should define implicit uncertainty about vehicle arrival. Any uncertainty should be covered either by the existing offset fields for travel time or the wait times described in GOFS Phase 1. The `*_window` is for the pickup/drop-off window at the administrative level, i.e. "I can access this service within this timeframe."

**Comment by westontrillium** [comment]
It looks like this issue can be closed now since the discussion evolved beyond this GH issue.
